pkg/maintainer/spv/moved_funds_sweep.go

```diff
@@ -0,0 +1,330 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to get block counter: [%v]", err)
+	}
+
+	currentBlock, err := blockCounter.CurrentBlock()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current block: [%v]", err)
+	}
+
+	// Calculate the starting block of the range in which the events will be
+	// searched for.
+	startBlock := currentBlock - historyDepth
+
+	events, err :=
+		spvChain.PastMovingFundsCommitmentSubmittedEvents(
+			&tbtc.MovingFundsCommitmentSubmittedEventFilter{
+				StartBlock: startBlock,
+			},
+		)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to get past moving funds commitment submitted events: [%v]",
+			err,
+		)
+	}
+
+	// Any wallet that was among target wallets recently could have created
+	// a moved funds sweep transaction.
+	targetWallets := [][20]byte{}
+	for _, event := range events {
+		targetWallets = append(targetWallets, event.TargetWallets...)
+	}
+
+	// Some target wallets may appear on the list multiple times. It can happen
+	// if multiple source wallets used the same target wallet. Make a list
+	// of unique wallets.
+	walletPublicKeyHashes := uniqueKeyHashes(targetWallets)
+
+	unprovenMovedFundsSweepTransactions := []*bitcoin.Transaction{}
+
+	for _, walletPublicKeyHash := range walletPublicKeyHashes {
+		wallet, err := spvChain.GetWallet(walletPublicKeyHash)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get wallet: [%v]", err)
+		}
+
+		if wallet.State != tbtc.StateLive &&
+			wallet.State != tbtc.StateMovingFunds {
+			// The wallet can only submit moved funds sweep proof if it's `Live`
+			//  or `MovingFunds`. If the state is different skip it.
+			continue
+		}
+
+		if wallet.PendingMovedFundsSweepRequestsCount == 0 {
+			// If the wallet does not have any pending moved funds sweep
+			// requests skip it.
+			continue
+		}
+
+		// When wallet makes a moved funds sweep transaction, it transfers
+		// funds to itself. Therefore we can search all the transactions that
+		// pay to the wallet's public key hash.
+		walletTransactions, err := btcChain.GetTransactionsForPublicKeyHash(
+			walletPublicKeyHash,
+			transactionLimit,
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to get transactions for wallet: [%v]",
+				err,
+			)
+		}
+
+		for _, transaction := range walletTransactions {
+			isUnproven, err :=
+				isUnprovenMovedFundsSweepTransaction(
+					transaction,
+					walletPublicKeyHash,
+					btcChain,
+					spvChain,
+				)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"failed to check if transaction is an unproven moved "+
+						"funds sweep transaction: [%v]",
+					err,
+				)
+			}
+
+			if isUnproven {
+				unprovenMovedFundsSweepTransactions = append(
+					unprovenMovedFundsSweepTransactions,
+					transaction,
+				)
+
+				// A wallet can have only one unproven moved funds sweep
+				// transaction at a time. If we found such transaction, we don't
+				// have to look at this wallet's transactions anymore.
+				break
```

Comment:
Just checking my understanding: Here we assume that the given wallet has always 0 or 1 unproven transaction (of any type) and `isUnprovenMovedFundsSweepTransaction` filters out previous moved funds sweep transactions that are already proven (using the `movedFundsSweepRequest.State != tbtc.MovedFundsStatePending` check). Is that correct?